### PR TITLE
[I2S][SR] Add new I2S library and Sound Recognition support

### PR DESCRIPTION
**Pull request by me-no-dev**
Redo of: https://github.com/espressif/arduino-esp32/pull/8617

### CODE REVIEW

File: libraries/ESP_I2S/src/ESP_I2S.h [hunk 1]
```diff
@@ -17,10 +27,16 @@ class I2SClass: public Stream {
...
     ~I2SClass();
 
+    //STD + TDM mode
     void setPins(int8_t bclk, int8_t ws, int8_t dout, int8_t din=-1, int8_t mclk=-1);
     void setInverted(bool bclk, bool ws, bool mclk=false);
 
-    bool begin(uint32_t rate, i2s_data_bit_width_t bits_cfg, i2s_slot_mode_t ch);
+    //PDM TX + PDM RX mode
+    void setPinsPdmTx(int8_t clk, int8_t dout0, int8_t dout1);
+    void setPinsPdmRx(int8_t clk, int8_t din0, int8_t din1, int8_t din2, int8_t din3);
```
Comment (me-no-dev) [inline libraries/ESP_I2S/src/ESP_I2S.h:36]: Have `dout1` and `din1+` default to `-1`. They are not always required. Or provide different signatures, depending on the number of channels.

File: libraries/ESP_I2S/src/ESP_I2S.cpp [hunk 2]
```diff
@@ -130,45 +232,229 @@ void I2SClass::setInverted(bool bclk, bool ws, bool mclk){
     _ws_inv = ws;
 }
 
-bool I2SClass::begin(uint32_t rate, i2s_data_bit_width_t bits_cfg, i2s_slot_mode_t ch){
+// Set pins for PDM TX mode
+void I2SClass::setPinsPdmTx(int8_t clk, int8_t dout0, int8_t dout1){
+    _tx_clk = clk;
+    _tx_dout0 = dout0;
+#if (SOC_I2S_PDM_MAX_TX_LINES > 1)
```
Comment (me-no-dev) [inline libraries/ESP_I2S/src/ESP_I2S.cpp:239]: See the other comment about method signatures.

File: libraries/ESP_I2S/src/ESP_I2S.cpp [hunk 3]
```diff
@@ -130,45 +232,229 @@ void I2SClass::setInverted(bool bclk, bool ws, bool mclk){
...
+                if(!perimanSetPinBus(_rx_din1, ESP32_BUS_TYPE_I2S_PDM, (void *)(this))){ goto err; }
+            }
+            if(_rx_din2 >= 0){
+                if(!perimanSetPinBus(_rx_din2, ESP32_BUS_TYPE_I2S_PDM, (void *)(this))){ goto err; }
+            }
+            if(_rx_din3 >= 0){
+                if(!perimanSetPinBus(_rx_din3, ESP32_BUS_TYPE_I2S_PDM, (void *)(this))){ goto err; }
+            }
+            break;
     }
```
Comment (me-no-dev) [inline libraries/ESP_I2S/src/ESP_I2S.cpp:441]: instead of having 4 cases one after another, have 4 separate private methods to init each type.

**Comment by me-no-dev** [comment]
Don't forget to fix the examples :)

File: libraries/ESP_I2S/src/ESP_I2S.h (hunk 1 above)
Comment (P-R-O-C-H-Y) [reply libraries/ESP_I2S/src/ESP_I2S.h]: done

File: libraries/ESP_I2S/src/ESP_I2S.cpp (hunk 3 above)
Comment (P-R-O-C-H-Y) [reply libraries/ESP_I2S/src/ESP_I2S.cpp:455]: Done. A lot cleaner code now :)

**Pull request merged**